airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/state_converter.py

```diff
@@ -117,27 +212,44 @@ def convert_from_sequential_state(self, stream_state: MutableMapping[str, Any])
             "legacy": stream_state,
         }
 
-    def convert_to_sequential_state(self, stream_state: MutableMapping[str, Any]) -> Any:
+    def convert_to_sequential_state(self, stream_state: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
         """
+        Convert the state message from the concurrency-compatible format to the stream's original format.
+
         e.g.
-        {
-            "state_type": "date-range",
-            "metadata": { … },
-            "slices": [
-                {starts: 0, end: 1617030403, finished_processing: true}
-            ]
-        }
-        =>
-        { "created": 1617030403 }
+        { "created": "2021-01-18T21:18:20.000Z" }
         """
         if self.is_state_message_compatible(stream_state):
             legacy_state = stream_state.get("legacy", {})
             if slices := stream_state.pop("slices", None):
-                legacy_state.update({self._cursor_field: self._get_latest_complete_time(slices)})
-            return legacy_state
+                legacy_state.update({self._cursor_field: self._millis_format(self._get_latest_complete_time(slices))})
+            return legacy_state or {}
         else:
             return stream_state
 
-    @staticmethod
-    def increment(timestamp: Any) -> Any:
-        return timestamp + 1
+    def _millis_format(self, timestamp: Optional[str]) -> Optional[str]:
+        return pendulum.parse(timestamp).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z" if timestamp else None
+
+    def _get_latest_complete_time(self, slices: List[MutableMapping[str, Any]]) -> Optional[str]:
+        """
+        Get the latest time before which all records have been processed.
+        """
+        if slices:
+            first_interval = self.merge_intervals(slices)[0][self.END_KEY]
+            return first_interval
+        else:
+            return None
+
+    def merge_intervals(self, intervals: List[MutableMapping[str, Any]]) -> List[MutableMapping[str, str]]:
```

Comment:
the merge_intervals implementations also look very similar to me. I expect them to converge into a private method in `ConcurrentStreamStateConverter` if the boundaries are always datetimes (or more likely into a yet-to-be-created `DatetimeStreamStateConverter`